### ISSUE TITLE
Factory contract unable to call deployed contract with --fork flag 

### ISSUE THREAD

**Issue by mds1** [trufflesuite/ganache-core#526]
[This discussion](https://forum.openzeppelin.com/t/issues-initializing-eip-1167-minimal-proxies-deployed-with-factory-contract/2039/7) on the OpenZeppelin forum contains a full description of the issue and an example that can be reproduced. Below I've included the core parts from that discussion.

The factory contract below should deploy a new `Forwarder` instance then call its `initialize()` function. The call to this function should emit the `Initialized` event and set `version` to 1.

This fails when testing against a forked version of the mainnnet, but succeeds otherwise. 

The error messages when testing against a forked version of the mainnet are below. From the stack trace shown, generated by testing against `ganache-cli -d -f https://mainnet.infura.io/v3/$INFURA_ID`, you can see from the "after each" section that the cause is a fork failure.

However, in my original project—before creating a trimmed down version to help diagnose this bug—this "after each" error did not show up in the stack trace, and I have not figured out why. 

Any ideas on the cause of this issue and how to fix it are greatly appreciated! Until it's resolved I don't really have a way to run some tests, since the full version of the `initialize()` function calls some functions on mainnet contracts. (There's always the option deploying and setting up local instances of all the dependent contracts, but I'm hoping to avoid that).

Thanks!

### Stack Trace

```
1) Contract: ForwarderFactory
       creates proxy:

      AssertionError: expected '0' to equal '1'
      + expected - actual

      -0
      +1
      
      at Context.it (test/forwarderFactory.test.js:30:55)
      at process._tickCallback (internal/process/next_tick.js:68:7)

  2) Contract: ForwarderFactory
       "after each" hook: after test for "creates proxy":
     Error: Returned error: Returned error: invalid argument 0: empty hex string
      at Object.ErrorResponse (node_modules/truffle/build/webpack:/node_modules/web3-core-helpers/src/errors.js:29:1)
      at /Users/mds/Documents/Documents-MBP/floatify/ganache-fork-failure/node_modules/truffle/build/webpack:/node_modules/web3-core-requestmanager/src/index.js:140:1
      at /Users/mds/Documents/Documents-MBP/floatify/ganache-fork-failure/node_modules/truffle/build/webpack:/packages/provider/wrapper.js:112:1
      at XMLHttpRequest.request.onreadystatechange (node_modules/truffle/build/webpack:/node_modules/web3-providers-http/src/index.js:96:1)
      at XMLHttpRequestEventTarget.dispatchEvent (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:1)
      at XMLHttpRequest._setReadyState (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:208:1)
      at XMLHttpRequest._onHttpResponseEnd (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:318:1)
      at IncomingMessage.<anonymous> (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:289:47)
      at endReadableNT (_stream_readable.js:1129:12)
      at process._tickCallback (internal/process/next_tick.js:63:19)
```

And here are the relevant files:

### FowarderFactory.sol

```solidity
pragma solidity ^0.5.0;

// Proxy factory contains the deployMinimal() function used below
// https://github.com/OpenZeppelin/openzeppelin-sdk/blob/v2.6.0/packages/lib/contracts/upgradeability/ProxyFactory.sol#L18
import "@openzeppelin/upgrades/contracts/upgradeability/ProxyFactory.sol";

import "./Forwarder.sol";

contract ForwarderFactory is ProxyFactory {

  address[] public forwarders;
  address[] public users;
  mapping (address => address) public getForwarder; // maps user => forwarder

  event ForwarderCreated(address indexed _address);

  function createForwarder(address _target, address _user) external {

    address _admin = msg.sender;
    bytes memory _payload = abi.encodeWithSignature("initialize(address,address)", _user, _admin);

    // Deploy proxy
    address _forwarder = deployMinimal(_target, _payload);
    emit ForwarderCreated(_forwarder);

    // Update state
    forwarders.push(_forwarder);
    users.push(_user);
    getForwarder[_user] = _forwarder;
  }
}
```
### Forwarder.sol

```solidity
pragma solidity ^0.5.0;

import "@openzeppelin/contracts-ethereum-package/contracts/ownership/Ownable.sol";
import "@openzeppelin/upgrades/contracts/Initializable.sol";

contract Forwarder is Initializable, Ownable {

  address public admin;
  uint256 public version;

  event Initialized(address indexed thisAddress);

  function initialize(address _recipient, address _admin) public initializer {
    emit Initialized(address(this));
    Ownable.initialize(_recipient);
    admin = _admin;
    version = 1;
  }
}
```

### forwarderFactory.test.js

```javascript
const {
  BN, // Big Number support
  constants, // Common constants, like the zero address and largest integers
  expectEvent, // Assertions for emitted events
  expectRevert, // Assertions for transactions that should fail
} = require('@openzeppelin/test-helpers');

const { expect } = require('chai');

const Forwarder = artifacts.require('Forwarder');
const ForwarderFactory = artifacts.require('ForwarderFactory');

contract('ForwarderFactory', ([admin, user]) => {
  it('creates proxy', async () => {
    const factory = await ForwarderFactory.new();
    const logic = await Forwarder.new();

    await factory.createForwarder(logic.address, user);

    const forwarderAddress = await factory.forwarders(0);
    const forwarder = await Forwarder.at(forwarderAddress);
    expect(await forwarder.version()).to.be.bignumber.equal('1');
  });
});
```

## Your Environment

* Version used:
* Environment name and version (e.g. PHP 5.4 on nginx 1.9.1):
* Server type and version:
* Operating System and version:
* Link to your project:

**Comment by nicholasjpaterno** [comment]
@mds1 Thank you for opening this issue!  We are looking into how we can address this ASAP!

**Comment by AcelisWeaven** [comment]
Hi @mds1 , by any chance, have you since found an alternative way to interact with mainnet contracts?

**Comment by mds1** [comment]
Unfortunately not. I ended up just deploying contracts to the mainnet to test the functionality in places where the ganache fork feature let me down.

Forking is pretty buggy, but when it works it's by far the easiest way to develop/test with other mainnet contracts. Really hope this feature gets some big updates soon 🤞

**Comment by AcelisWeaven** [comment]
Oh that's a bummer! Thanks for the quick reply.

Yeah that definitely seems to be the easiest way. I'm hoping for an update too 👍

**Comment by AcelisWeaven** [comment]
@mds1 Sorry for the double post, just wanted to let you know that I successfully emulated the same feature. The process is not as simple as `ganache-cli` sadly, but it allows contracts to be tested on mainnet, and they can even be debugged on Remix.

1) Setup a `geth` full node
2) Optional, but recommended: Copy the datadir to another place. I'll refer to the new directory as `<new-data-dir>` below.
3) Recompile `geth` to have no difficulty to mine blocks. -> https://hackernoon.com/how-to-reduce-block-difficulty-in-ethereum-private-testnet-2ad505609e82
4) Run your modified `geth`. Here's my command:
```sh
my-geth.exe --rpc --rpcapi=eth,web3,net,personal --rpccorsdomain "*" --datadir <new-data-dir> --maxpeers 0 --nat none --nodiscover --v5disc=false --port 0 --allow-insecure-unlock --unlock <your-node-address> --mine --miner.threads=1 --networkid=2020
```

Everything seems to work fine! Sadly, you'll miss out on `truffle` features, but it's a temporary solution.

**Comment by mds1** [comment]
@AcelisWeaven This is great to know, thanks for sharing. I'd love to see this pulled out into an article/tutorial, with some details explaining the different CLI flags for the geth commands. I'm sure a lot of people would find it valuable!

**Comment by mds1** [comment]
So I'm once again getting the same error as above, `Error: Returned error: Returned error: invalid argument 0: empty hex string`.

But this time, it has nothing to do with factory contracts. [Here](https://github.com/mds1/Bancor-Fiat-On-Ramp/tree/4b734c35d3e661c0d284df029e53e6b22ea12a05) is a link to the specific commit where it currently fails with that error so you can reproduce this.

I cannot figure out what is causing this error, so any solutions or workaround are greatly appreciated! (Hoping not to have to setup my own geth node 😁)

I'm using Ganache CLI v6.9.1 (ganache-core: 2.10.2).

**Comment by seesemichaelj** [comment]
@mds1 Hey! I'm looking at this and other forking bugs now. Am I correct that you were deploying **both** the ForwarderFactory and Forwarder contracts post fork? I'm looking over these tests, and it doesn't look like you're referring to a contract deployed on Mainnet at all. Is that correct?

**Comment by mds1** [comment]
Hey @seesemichaelj, super excited to see that you are looking into these forking bugs!!

Yes, you are correct that both the ForwarderFactory and Forwarder contracts were deployed after forking, i.e. they were both deployed on a local ganache instance that forked that mainnet

**Comment by seesemichaelj** [comment]
Cool, thanks @mds1! You don't happen to know what version of `ganache-cli` you were using back then, do you?

**Comment by seesemichaelj** [comment]
Oh, looks like you mentioned it [on your other post in the OpenZeppelin thread](https://forum.openzeppelin.com/t/issues-initializing-eip-1167-minimal-proxies-deployed-with-factory-contract/2039): ganache-cli v6.6.0, ganache-core v2.7.0

**Comment by mds1** [comment]
Ah good find, it seems I left that out of the description here 😬

From the March 24 comment above it seems I was getting the same error message, but with a different cause, using ganache-cli v6.9.1 (ganache-core: 2.10.2)

**Comment by seesemichaelj** [comment]
Just for bookkeeping: I am able to reproduce the original error (ForwarderFactory test failing) with `ganache-cli@6.6.0`, but not `ganache@6.9.1+` (I haven't yet determined the specific version it got fixed in). I also reproduce only seeing the first error (`expected 0 to equal 1`) but not the second one (`invalid argument 0: empty hex string`)

**Comment by seesemichaelj** [comment]
@mds1 I'm not reproducing the Bancor repo bug with `truffle test` using `ganache-cli@6.9.1` forked off of Mainnet. I get this error:

```
$ truffle test

Compiling your contracts...
===========================
✔ Fetching solc version list from solc-bin. Attempt #1
> Compiling ./contracts/IBancorConverter.sol
> Compiling ./contracts/ProvideLiquidity.sol
> Compiling @openzeppelin/contracts/token/ERC20/IERC20.sol
✔ Fetching solc version list from solc-bin. Attempt #1
> Compilation warnings encountered:

    /home/mike/work/truffle/Bancor-Fiat-On-Ramp/contracts/IBancorConverter.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
,/home/mike/work/truffle/Bancor-Fiat-On-Ramp/contracts/ProvideLiquidity.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
,@openzeppelin/contracts/token/ERC20/IERC20.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

> Artifacts written to /tmp/test--196107-TaiNU40YD16d
> Compiled successfully using:
   - solc: 0.6.12+commit.27d51765.Emscripten.clang



  Contract: Provide Liquidity
    1) "before each" hook for "deploys properly"


  0 passing (5s)
  1 failing

  1) Contract: Provide Liquidity
       "before each" hook for "deploys properly":
     Error: Returned error: sender account not recognized
      at Context.beforeEach (test/provideLiquidity.js:36:23)
      at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by mds1** [comment]
@seesemichaelj Hmm, let me look into the Bancor one this week and see if memory helps me reproduce it.

I'm currently working on a new project which involves a nearly identical architecture of a factory deploying EIP-1167 minimal proxies, and that will also require a forked mainnet to test. I expect to be running tests for that this week, so will let you know if these errors (or any others) show up then

**Comment by seesemichaelj** [comment]
Thanks!

**Comment by seesemichaelj** [comment]
Hey @mds1, were you able to reproduce these errors?

**Comment by mds1** [comment]
Ok so I just checked out the Bancor commit ([this comment above](https://github.com/trufflesuite/ganache-core/issues/526#issuecomment-603571459)) and was able to reproduce that one using Ganache CLI v6.9.1 (ganache-core: 2.10.2). These steps should let you reproduce:
1. In `test/provideLiquidity.js`, comment out line 37: `expect((await DaiContract.balanceOf(alice)).toString()).to.equal(amount);`
2. In `test/provideLiquidity.js`, uncomment out all tests
3. Make sure Ganache CLI v6.9.1 is installed globally
4. Install packages with `yarn`
5. Set up the `.env` file at the project root as described in the README
6. Start ganache with `yarn run ganache`
7. Run tests with `yarn test`

And the full output is below:

```
  Contract: Provide Liquidity
2
5
    ✓ deploys properly
    1) lets users provide liquidity
    2) "after each" hook: after test for "lets users provide liquidity"


  1 passing (12s)
  2 failing

  1) Contract: Provide Liquidity
       lets users provide liquidity:
     Error: Returned error: VM Exception while processing transaction: revert
     at PromiEvent (node_modules/truffle/build/webpack:/packages/contract/lib/promievent.js:9:1)
      at TruffleContract.provideLiquidity (node_modules/truffle/build/webpack:/packages/contract/lib/execute.js:169:1)
      at Context.<anonymous> (test/provideLiquidity.js:54:36)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)

  2) Contract: Provide Liquidity
       "after each" hook: after test for "lets users provide liquidity":
     Error: Returned error: Returned error: invalid argument 0: empty hex string
      at Object.ErrorResponse (node_modules/truffle/build/webpack:/node_modules/web3-core-helpers/src/errors.js:29:1)
      at /Users/mds/Documents/projects/Bancor-Fiat-On-Ramp/node_modules/truffle/build/webpack:/node_modules/web3-core-requestmanager/src/index.js:140:1
      at /Users/mds/Documents/projects/Bancor-Fiat-On-Ramp/node_modules/truffle/build/webpack:/packages/provider/wrapper.js:112:1
      at XMLHttpRequest.request.onreadystatechange (node_modules/truffle/build/webpack:/node_modules/web3-providers-http/src/index.js:96:1)
      at XMLHttpRequestEventTarget.dispatchEvent (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:1)
      at XMLHttpRequest._setReadyState (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:208:1)
      at XMLHttpRequest._onHttpResponseEnd (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:318:1)
      at IncomingMessage.<anonymous> (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:289:47)
      at endReadableNT (_stream_readable.js:1220:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

The first error on line 54 is a valid revert, but I'm not sure why the `after` hooks throws an `invalid argument 0` error, because those arguments are valid. It seems to be related to the revert because if you comment out that test (line 54) the `after` hook runs without issue. In both cases (running and not running the test that reverts), no balances or state are changing, so the `after` hook should run successfully in both cases.

I tried this with Ganache CLI v6.10.2 (ganache-core: 2.11.3) and get the same output

**Comment by seesemichaelj** [comment]
Heads up @mds1, I'm not sure if the `v2.12.2-beta.0` fixes this or not. I still need to get back to forking stuff as I've been doing some Truffle Teams stuff. It's possible something in the beta release fixed this issue, but it wouldn't have been on purpose haha. We did some rearchitecture of the forking storage handling, so it's possible things are better now

I'll hopefully give this a shot sometime soon!

**Comment by seesemichaelj** [comment]
I'm able to reproduce your errors and `ganache-cli@6.11.1-alpha.1` (which includes my recent fixes) doesn't seem to remove them.

I'm not sure how this error is related to "factory contract unable to call deployed contract" since it sounds like we're having issues calling the DaiContract given the fact that we've made some other call.

I'll keep digging through this issue, but I'm not sure how it's related to factory contracts

**Comment by seesemichaelj** [comment]
Also, do you know why `bancorConverter.convert2` reverts (since you say it's a valid revert)? This may help me debug

**Comment by seesemichaelj** [comment]
@mds1 upgrading `truffle` from `5.1.18` to `5.1.47` seemed to resolve the issue for me.

I'm not sure of the root cause, but the error was happening from `ganache-core`/web3 because some package in the chain was sending `eth_getBlockByNumber` with params `[null, true]` where we/the spec expects a quantity hex string

I'm guessing there was a bug in truffle fixed sometime since March that fixed this issue (or it could have even been an update on web3). I'll spend a little time to see if I can find the PR which fixed the bug, but let me know if that also fixes the issue for you

**Comment by seesemichaelj** [comment]
It seems that the issue gets resolved in `truffle@5.1.20`. However, I've scoured the PRs an [commits](https://github.com/trufflesuite/truffle/compare/v5.1.19...v5.1.20) that went into 5.1.20, and I can't see anything jumping out that would have fixed this issue 🤷 I'm at a loss on this one, other than it's working for me after upgradeing to 5.1.20+

**Comment by mds1** [comment]
> Also, do you know why bancorConverter.convert2 reverts (since you say it's a valid revert)? This may help me debug

Unfortunately I don't. I did some digging and can't remember what the bug was here, but it seems I resolved this demo by just switching to an ETH-token conversion instead of a token-token conversion. So perhaps the bug was related to ERC20 approvals, but I'm not certain of that.

However, I can confirm that upgrading truffle to ^5.1.47 resolves the issue!
